specutils/manipulation/resample.py

```diff
@@ -0,0 +1,189 @@
+from abc import ABC, abstractmethod
+
+import numpy as np
+from astropy.units import Quantity
+
+from ..spectra import Spectrum1D
+
+__all__ = ['ResampleBase','FluxConservingResample']
+
+
+class ResampleBase(ABC):
+    """
+    Base class for resample classes.  The algorithms and needs for difference
+    resamples will vary quite a bit, so this class is relatively sparse.
+    """
+
+    @abstractmethod
+    def __call__(self, orig_spectrum, fin_lamb):
+        """
+        Return the resulting `~specutils.Spectrum1D` of the resampling.
+        """
+        return NotImplemented
+
+    @abstractmethod
+    def resample(self, orig_spectrum, fin_lamb, weights='unc'):
+        """
+        Workhorse method that will return the resampled Spectrum1D
+        object.
+        """
+        return NotImplemented
+
+    @staticmethod
+    def _bin_edges(x):
+        """
+        Calculate the bin edge values of an input dispersion axis. Input values
+        are assumed to be the center of the bins.
+
+        todo: this should live in the main spectrum object, but we're still
```

Comment:
The main `Spectrum1D` object already contains a `bin_edges` property.

Reply:
I tried to use this and it didn't seem to return anything, will look into it again.

Reply:
It might be that you're looking to return the dispersion grid shifted for bin edges, the the `bin_edges` property also performed the `pixel_to_world` transformation after calculating the edges.

Reply:
This is what I get.... i have the latest version of the gwcs library...

```
In [5]: flux_val = np.array([1, 3, 7, 6, 20])                                                                                                                                   

In [6]: wave_val = np.array([2, 4, 12, 16, 20])                                                                                                                                 

In [7]: input_spectra = Spectrum1D(flux_val*u.mJy, wave_val* u.AA)                                                                                                              

In [8]: input_spectra.bin_edges                                                                                                                                                 
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-8-c112932b4a4f> in <module>
----> 1 input_spectra.bin_edges

~/miniconda3/envs/specutilsdev/lib/python3.7/site-packages/astropy/utils/decorators.py in __get__(self, obj, owner)
    724                 return val
    725             else:
--> 726                 val = self.fget(obj)
    727                 obj.__dict__[self._key] = val
    728                 return val

~/specutils/specutils/spectra/spectrum1d.py in bin_edges(self)
    148     @lazyproperty
    149     def bin_edges(self):
--> 150         return self.wcs.bin_edges()
    151 
    152     @property

AttributeError: 'GWCSAdapter' object has no attribute 'bin_edges'
```

Reply:
Right, well, it seems the bin edges are only calculated for fits wcs objects. :man_shrugging: